product_multi_relation/models/product_relation_all.py

```diff
@@ -0,0 +1,471 @@
+# Copyright 2014-2018 Therp BV <http://therp.nl>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+# pylint: disable=method-required-super
+import collections
+import logging
+
+from psycopg2.extensions import AsIs
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+from odoo.tools import drop_view_if_exists
+
+
+_logger = logging.getLogger(__name__)
+
+
+# Register relations
+RELATIONS_SQL = """\
+SELECT
+    (rel.id * %%(padding)s) + %(key_offset)s AS id,
+    'product.relation' AS res_model,
+    rel.id AS res_id,
+    rel.left_product_id AS this_product_id,
+    rel.right_product_id AS other_product_id,
+    rel.type_id,
+    rel.date_start,
+    rel.date_end,
+    %(is_inverse)s as is_inverse
+    %(extra_additional_columns)s
+FROM product_relation rel"""
+
+# Register inverse relations
+RELATIONS_SQL_INVERSE = """\
+SELECT
+    (rel.id * %%(padding)s) + %(key_offset)s AS id,
+    'product.relation',
+    rel.id,
+    rel.right_product_id,
+    rel.left_product_id,
+    rel.type_id,
+    rel.date_start,
+    rel.date_end,
+    %(is_inverse)s as is_inverse
+    %(extra_additional_columns)s
+FROM product_relation rel"""
+
+
+class ProductRelationAll(models.AbstractModel):
+    """Abstract model to show each relation from two sides."""
+    _auto = False
+    _log_access = False
+    _name = 'product.relation.all'
+    _description = 'All (non-inverse + inverse) relations between products'
+    _order = \
+        'this_product_id, type_selection_id, date_end desc, date_start desc'
+
+    res_model = fields.Char(
```

Comment:
Is there a use case where res_model cannot be a product ?

Reply:
Not sure I follow here.

Reply:
What can represent res_model/res_id ? Reading the code, it appears to me it can be only product.template. Am I wrong ?

Reply:
You're most probably right. But it eludes me what you're trying to say here. Is it that because it can only be product.template the field is not really needed?
Please understand, that what I did here is adapted the code from original module (partner_multi_contact) to work with product.template instead and also updated demo/test data so that they make some sense. As I am quite a beginner in Python/Odoo, I do not fully understand some parts of the code and only guess. This certainly could be one of these parts :-)

Reply:
Nope, what I want to say is that in that case, you should simply replace both fields by a Many2one.